File: behavioral/dart/strategy_structured/route_planning.dart
import 'interfaces.dart';

class FastestRoute implements RouteStrategy {
  @override
  void calculateRoute(String from, String to) {
    print('Calculating fastest route from $from to $to');
    print('  Using highways and main roads to minimize time');
    print('  Estimated time: 25 minutes');
  }
}

class ShortestRoute implements RouteStrategy {
  @override
  void calculateRoute(String from, String to) {
    print('Calculating shortest route from $from to $to');
    print('  Using direct path to minimize distance');
    print('  Estimated distance: 15 km');
  }
}

class ScenicRoute implements RouteStrategy {
  @override
  void calculateRoute(String from, String to) {
    print('Calculating scenic route from $from to $to');
    print('  Using parks and landmarks for best views');
    print('  Estimated time: 45 minutes');
  }
}

class NavigationSystem {
  RouteStrategy? _strategy;
  
  void setRouteStrategy(RouteStrategy strategy) {
    _strategy = strategy;
  }
  
  void navigate(String from, String to) {
    print('\n=== Navigation ===');
    print('From: $from');
    print('To: $to');
    
    if (_strategy != null) {
      _strategy!.calculateRoute(from, to);
    } else {
      print('No route strategy selected');
    }
  }
}

File: behavioral/typescript/strategy_structured/route_planning.ts
import { RouteStrategy } from './interfaces';

export class FastestRoute implements RouteStrategy {
  calculateRoute(from: string, to: string): void {
    console.log(`Calculating fastest route from ${from} to ${to}`);
    console.log('  Using highways and main roads to minimize time');
    console.log('  Estimated time: 25 minutes');
  }
}

export class ShortestRoute implements RouteStrategy {
  calculateRoute(from: string, to: string): void {
    console.log(`Calculating shortest route from ${from} to ${to}`);
    console.log('  Using direct path to minimize distance');
    console.log('  Estimated distance: 15 km');
  }
}

export class ScenicRoute implements RouteStrategy {
  calculateRoute(from: string, to: string): void {
    console.log(`Calculating scenic route from ${from} to ${to}`);
    console.log('  Using parks and landmarks for best views');
    console.log('  Estimated time: 45 minutes');
  }
}

export class NavigationSystem {
  private strategy: RouteStrategy | null = null;
  
  setRouteStrategy(strategy: RouteStrategy): void {
    this.strategy = strategy;
  }
  
  navigate(from: string, to: string): void {
    console.log('\n=== Navigation ===');
    console.log(`From: ${from}`);
    console.log(`To: ${to}`);
    
    if (this.strategy) {
      this.strategy.calculateRoute(from, to);
    } else {
      console.log('No route strategy selected');
    }
  }
}
